File: src/features/notifications/scheduler.ts
import * as Notifications from 'expo-notifications';
import * as Device from 'expo-device';
import { Platform } from 'react-native';
import i18n from '@/i18n';
import { listActiveMedicationsWithSchedule } from '@/db/repositories/prescriptions';
import { getBoolSetting, SettingsKeys } from '@/db/repositories/settings';
import { dayjs } from '@/lib/date';

const ANDROID_CHANNEL_ID = 'medication-reminders';

export interface ReminderPrefs {
  sound: boolean;
  vibration: boolean;
}

export async function getReminderPrefs(): Promise<ReminderPrefs> {
  return {
    sound: await getBoolSetting(SettingsKeys.reminderSound, true),
    vibration: await getBoolSetting(SettingsKeys.reminderVibration, true),
  };
}

// Foreground display behaviour
Notifications.setNotificationHandler({
  handleNotification: async () => ({
    shouldShowBanner: true,
    shouldShowList: true,
    shouldPlaySound: true,
    shouldSetBadge: true,
  }),
});

export async function configureNotifications(): Promise<void> {
  if (Platform.OS === 'android') {
    const prefs = await getReminderPrefs();
    await Notifications.setNotificationChannelAsync(ANDROID_CHANNEL_ID, {
      name: i18n.t('reminders.channelName'),
      importance: Notifications.AndroidImportance.HIGH,
      sound: prefs.sound ? 'default' : undefined,
      // Android channel vibration is immutable once created; an empty pattern
      // disables it. Recreated via applyReminderPrefs() when the user toggles.
      vibrationPattern: prefs.vibration ? [0, 250, 250, 250] : [0],
      enableVibrate: prefs.vibration,
      lightColor: '#0E7C7B',
      lockscreenVisibility: Notifications.AndroidNotificationVisibility.PUBLIC,
    });
  }
}

/**
 * Re-applies reminder preferences: recreates the Android channel (channels are
 * immutable, so we delete + recreate) and reschedules all reminders so the new
 * sound setting takes effect on already-queued notifications.
 */
export async function applyReminderPrefs(patientId: number | null): Promise<void> {
  if (Platform.OS === 'android') {
    await Notifications.deleteNotificationChannelAsync(ANDROID_CHANNEL_ID).catch(() => {});
    await configureNotifications();
  }
  if (patientId != null) await syncReminders(patientId);
}

export async function requestNotificationPermission(): Promise<boolean> {
  if (!Device.isDevice) return false;
  const settings = await Notifications.getPermissionsAsync();
  let granted = settings.granted || settings.ios?.status === Notifications.IosAuthorizationStatus.PROVISIONAL;
  if (!granted) {
    const req = await Notifications.requestPermissionsAsync({
      ios: { allowAlert: true, allowBadge: true, allowSound: true },
    });
    granted = req.granted;
  }
  return granted;
}

export async function hasNotificationPermission(): Promise<boolean> {
  const settings = await Notifications.getPermissionsAsync();
  return settings.granted;
}

interface ScheduleArgs {
  medicationId: number;
  medicationName: string;
  dosage: string | null;
  hour: number;
  minute: number;
  sound: boolean;
}

async function scheduleDailyDose({
  medicationId,
  medicationName,
  dosage,
  hour,
  minute,
  sound,
}: ScheduleArgs): Promise<void> {
  await Notifications.scheduleNotificationAsync({
    content: {
      title: i18n.t('reminders.doseTitle'),
      body: i18n.t('reminders.doseBody', {
        medication: medicationName,
        dosage: dosage ?? '',
      }),
      sound: sound ? 'default' : undefined,
      data: { type: 'dose', medicationId },
      ...(Platform.OS === 'android' ? { channelId: ANDROID_CHANNEL_ID } : {}),
    },
    trigger: {
      type: Notifications.SchedulableTriggerInputTypes.DAILY,
      hour,
      minute,
    },
  });
}

/**
 * Cancel all scheduled notifications and reschedule daily reminders for every
 * active medication / schedule time. Call after any prescription change.
 */
export async function syncReminders(patientId: number): Promise<void> {
  const granted = await hasNotificationPermission();
  if (!granted) return;

  await Notifications.cancelAllScheduledNotificationsAsync();

  const prefs = await getReminderPrefs();
  const meds = await listActiveMedicationsWithSchedule(patientId);
  for (const { medication, times } of meds) {
    // Skip medications that have ended
    if (medication.startDate && medication.durationDays) {
      const end = dayjs(medication.startDate).add(medication.durationDays - 1, 'day').endOf('day');
      if (dayjs().isAfter(end)) continue;
    }
    for (const t of times) {
      const [hour, minute] = t.time.split(':').map(Number);
      await scheduleDailyDose({
        medicationId: medication.id,
        medicationName: medication.name,
        dosage: medication.dosage,
        hour,
        minute,
        sound: prefs.sound,
      });
    }
  }
}

export async function scheduleAppointmentReminder(
  title: string,
  body: string,
  when: Date,
): Promise<string | null> {
  if (dayjs(when).isBefore(dayjs())) return null;
  return Notifications.scheduleNotificationAsync({
    content: {
      title,
      body,
      sound: 'default',
      ...(Platform.OS === 'android' ? { channelId: ANDROID_CHANNEL_ID } : {}),
    },
    trigger: {
      type: Notifications.SchedulableTriggerInputTypes.DATE,
      date: when,
    },
  });
}

export async function cancelNotification(id: string): Promise<void> {
  await Notifications.cancelScheduledNotificationAsync(id);
}

File: lib/features/notifications/scheduler.dart
import 'dart:io' show Platform;
import 'dart:ui' show Color;

import 'package:flutter_local_notifications/flutter_local_notifications.dart';
import 'package:flutter_timezone/flutter_timezone.dart';
import 'package:timezone/data/latest_all.dart' as tzdata;
import 'package:timezone/timezone.dart' as tz;

import '../../db/repositories/prescriptions_repository.dart';
import '../../db/repositories/settings_repository.dart';
import '../../i18n/app_localizations.dart';

/// Local dose reminders. Ported from `src/features/notifications/scheduler.ts`
/// (expo-notifications → flutter_local_notifications).
///
/// Reminders repeat daily at a wall-clock time, so they are scheduled in the
/// device's timezone rather than UTC: a dose set for 08:00 must stay 08:00 when
/// the user travels.

const String androidChannelId = 'medication-reminders';

class ReminderPrefs {
  final bool sound;
  final bool vibration;

  const ReminderPrefs({required this.sound, required this.vibration});
}

class NotificationScheduler {
  NotificationScheduler({
    FlutterLocalNotificationsPlugin? plugin,
    this.settings = const SettingsRepository(),
    this.prescriptions = const PrescriptionsRepository(),
  }) : _plugin = plugin ?? FlutterLocalNotificationsPlugin();

  final FlutterLocalNotificationsPlugin _plugin;
  final SettingsRepository settings;
  final PrescriptionsRepository prescriptions;

  bool _timezoneReady = false;

  Future<ReminderPrefs> getReminderPrefs() async => ReminderPrefs(
        sound: await settings.getBool(SettingsKeys.reminderSound, true),
        vibration: await settings.getBool(SettingsKeys.reminderVibration, true),
      );

  Future<void> _ensureTimezone() async {
    if (_timezoneReady) return;
    tzdata.initializeTimeZones();
    try {
      tz.setLocalLocation(tz.getLocation(await FlutterTimezone.getLocalTimezone()));
    } catch (_) {
      // Falling back to UTC would shift every reminder; keep the package
      // default local zone instead of guessing.
    }
    _timezoneReady = true;
  }

  /// Must run before any scheduling. Safe to call more than once.
  Future<void> initialize(Translations t) async {
    await _ensureTimezone();

    const androidInit = AndroidInitializationSettings('@mipmap/ic_launcher');
    const darwinInit = DarwinInitializationSettings(
      // Permission is requested explicitly later, in context, so the prompt
      // appears when the user has a reason to say yes.
      requestAlertPermission: false,
      requestBadgePermission: false,
      requestSoundPermission: false,
    );
    await _plugin.initialize(
      const InitializationSettings(android: androidInit, iOS: darwinInit),
    );

    await configureAndroidChannel(t);
  }

  Future<void> configureAndroidChannel(Translations t) async {
    if (!Platform.isAndroid) return;
    final prefs = await getReminderPrefs();
    final android = _plugin.resolvePlatformSpecificImplementation<
        AndroidFlutterLocalNotificationsPlugin>();
    await android?.createNotificationChannel(
      AndroidNotificationChannel(
        androidChannelId,
        t.t('reminders.channelName'),
        importance: Importance.high,
        playSound: prefs.sound,
        enableVibration: prefs.vibration,
        ledColor: const Color(0xFF0E7C7B),
      ),
    );
  }

  /// Re-applies reminder preferences. Android channels are immutable once
  /// created, so the channel is deleted and recreated, then reminders are
  /// rescheduled so queued notifications pick up the new sound setting.
  Future<void> applyReminderPrefs(int? patientId, Translations t) async {
    if (Platform.isAndroid) {
      final android = _plugin.resolvePlatformSpecificImplementation<
          AndroidFlutterLocalNotificationsPlugin>();
      try {
        await android?.deleteNotificationChannel(androidChannelId);
      } catch (_) {
        // Nothing to delete on a fresh install.
      }
      await configureAndroidChannel(t);
    }
    if (patientId != null) await syncReminders(patientId, t);
  }

  Future<bool> requestPermission() async {
    if (Platform.isIOS) {
      final ios = _plugin.resolvePlatformSpecificImplementation<
          IOSFlutterLocalNotificationsPlugin>();
      final granted = await ios?.requestPermissions(
        alert: true,
        badge: true,
        sound: true,
      );
      return granted ?? false;
    }
    if (Platform.isAndroid) {
      final android = _plugin.resolvePlatformSpecificImplementation<
          AndroidFlutterLocalNotificationsPlugin>();
      final granted = await android?.requestNotificationsPermission();
      return granted ?? false;
    }
    return false;
  }

  Future<bool> hasPermission() async {
    if (Platform.isAndroid) {
      final android = _plugin.resolvePlatformSpecificImplementation<
          AndroidFlutterLocalNotificationsPlugin>();
      return await android?.areNotificationsEnabled() ?? false;
    }
    if (Platform.isIOS) {
      final ios = _plugin.resolvePlatformSpecificImplementation<
          IOSFlutterLocalNotificationsPlugin>();
      final settings = await ios?.checkPermissions();
      return settings?.isAlertEnabled ?? false;
    }
    return false;
  }

  /// Stable per-(medication, time) id so rescheduling replaces rather than
  /// duplicates. Kept inside 32 bits, which is the platform id limit.
  static int notificationId(int medicationId, int hour, int minute) =>
      ((medicationId * 1440) + (hour * 60) + minute) & 0x7FFFFFFF;

  NotificationDetails _details(ReminderPrefs prefs, String channelName) {
    return NotificationDetails(
      android: AndroidNotificationDetails(
        androidChannelId,
        channelName,
        importance: Importance.high,
        priority: Priority.high,
        playSound: prefs.sound,
        enableVibration: prefs.vibration,
      ),
      iOS: DarwinNotificationDetails(presentSound: prefs.sound),
    );
  }

  tz.TZDateTime _nextInstanceOf(int hour, int minute) {
    final now = tz.TZDateTime.now(tz.local);
    var scheduled =
        tz.TZDateTime(tz.local, now.year, now.month, now.day, hour, minute);
    if (!scheduled.isAfter(now)) {
      scheduled = scheduled.add(const Duration(days: 1));
    }
    return scheduled;
  }

  /// Cancels every scheduled reminder and reschedules daily ones for each
  /// active medication / schedule time. Call after any prescription change.
  ///
  /// Returns false when notifications are not permitted — the caller decides
  /// whether to warn, because silently doing nothing is how users end up with
  /// medications and no reminders.
  Future<bool> syncReminders(int patientId, Translations t) async {
    if (!await hasPermission()) return false;
    await _ensureTimezone();
    await _plugin.cancelAll();

    final prefs = await getReminderPrefs();
    final details = _details(prefs, t.t('reminders.channelName'));
    final meds = await prescriptions.listActiveMedicationsWithSchedule(patientId);
    final today = DateTime.now();

    for (final entry in meds) {
      final med = entry.medication;

      // Skip medications whose course has already ended.
      if (med.startDate != null && med.durationDays != null) {
        final start = DateTime.tryParse(med.startDate!);
        if (start != null && med.durationDays! > 0) {
          final end = DateTime(start.year, start.month, start.day)
              .add(Duration(days: med.durationDays! - 1))
              .add(const Duration(hours: 23, minutes: 59));
          if (today.isAfter(end)) continue;
        }
      }

      for (final time in entry.times) {
        final parts = time.time.split(':');
        final hour = int.tryParse(parts.isNotEmpty ? parts[0] : '');
        final minute = int.tryParse(parts.length > 1 ? parts[1] : '');
        if (hour == null || minute == null) continue;

        await _plugin.zonedSchedule(
          notificationId(med.id, hour, minute),
          t.t('reminders.doseTitle'),
          t.t('reminders.doseBody', params: {
            'medication': med.name,
            'dosage': med.dosage ?? '',
          }),
          _nextInstanceOf(hour, minute),
          details,
          androidScheduleMode: AndroidScheduleMode.exactAllowWhileIdle,
          // Wall-clock, not absolute: an 08:00 dose stays 08:00 across DST
          // and timezone changes.
          uiLocalNotificationDateInterpretation:
              UILocalNotificationDateInterpretation.wallClockTime,
          matchDateTimeComponents: DateTimeComponents.time, // repeat daily
        );
      }
    }
    return true;
  }

  Future<void> scheduleAppointmentReminder({
    required int id,
    required String title,
    required String body,
    required DateTime when,
    required Translations t,
  }) async {
    await _ensureTimezone();
    final prefs = await getReminderPrefs();
    await _plugin.zonedSchedule(
      id,
      title,
      body,
      tz.TZDateTime.from(when, tz.local),
      _details(prefs, t.t('reminders.channelName')),
      androidScheduleMode: AndroidScheduleMode.exactAllowWhileIdle,
      uiLocalNotificationDateInterpretation:
          UILocalNotificationDateInterpretation.wallClockTime,
    );
  }

  Future<void> cancel(int id) => _plugin.cancel(id);

  Future<void> cancelAll() => _plugin.cancelAll();
}
